src/prefect/environments/storage/s3.py

```diff
@@ -0,0 +1,154 @@
+
+    Args:
+        - aws_access_key_id (str, optional): AWS access key id for connecting to S3.
+            Defaults to the value set in the environment variable
+            `AWS_ACCESS_KEY_ID` or `None`
+        - aws_secret_access_key (str, optional): AWS secret access key for connecting to S3.
+            Defaults to the value set in the environment variable
+            `AWS_SECRET_ACCESS_KEY` or `None`
+        - aws_session_token (str, optional): AWS session key for connecting to S3
+            Defaults to the value set in the environment variable
+            `AWS_SESSION_TOKEN` or `None`
+        - bucket (str, optional): the name of the S3 Bucket to store the Flow
+        - key (str, optional): a unique key to use for uploading this Flow to S3
+    """
+
+    def __init__(
+        self,
+        aws_access_key_id: str = None,
+        aws_secret_access_key: str = None,
+        aws_session_token: str = None,
+        bucket: str = None,
+        key: str = None,
+    ) -> None:
+        self.flows = dict()  # type: Dict[str, str]
+        self.bucket = bucket or ""
+        self.key = key
+
+        self.aws_access_key_id = aws_access_key_id
+        self.aws_secret_access_key = aws_secret_access_key
+        self.aws_session_token = aws_session_token
+
+        super().__init__()
+
+    def get_flow(self, flow_location: str) -> "Flow":
+        """
+        Given a flow_location within this Storage object, returns the underlying Flow (if possible).
+
+        Args:
+            - flow_location (str): the location of a flow within this Storage; in this case,
+                a file path where a Flow has been serialized to
+
+        Returns:
+            - Flow: the requested flow
+
+        Raises:
+            - ValueError: if the flow is not contained in this storage
+        """
+        if not flow_location in self.flows.values():
+            raise ValueError("Flow is not contained in this Storage")
+
+        stream = io.BytesIO()
+
+        # Download stream from S3
+        self._boto3_client.download_fileobj(
+            Bucket=self.bucket, Key=flow_location, Fileobj=stream
+        )
+
+        ## prepare data and return
+        stream.seek(0)
+        output = stream.read()
+
+        return cloudpickle.loads(output)
+
+    def add_flow(self, flow: "Flow") -> str:
+        """
+        Method for storing a new flow as bytes in the local filesytem.
+
+        Args:
+            - flow (Flow): a Prefect Flow to add
+
+        Returns:
+            - str: the location of the newly added flow in this Storage object
+
+        Raises:
+            - ValueError: if a flow with the same name is already contained in this storage
+        """
+        if flow.name in self:
+            raise ValueError(
+                'Name conflict: Flow with the name "{}" is already present in this storage.'.format(
+                    flow.name
+                )
+            )
+
+        # Pickle Flow
```

Comment:
So, purely following our docker pattern, we might want to consider actually performing the upload at `build` time, although I see how that conflicts with how we're storing `self.flows` here - thoughts?

Reply:
Yeah we should definitely move it to `build` time, good call

Reply:
Bump on the above comment @joshmeek